File: src/main/java/de/com/fdm/core/RedisListener.java
package de.com.fdm.core;

import de.com.fdm.tmi.Reader;
import org.redisson.Redisson;
import org.redisson.api.RTopic;
import org.redisson.api.RedissonClient;
import org.redisson.config.Config;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.IOException;

@Service
public class RedisListener {

    @Autowired
    public RedisListener(Reader reader) throws IOException {
        Config config = Config.fromYAML(new File("src/main/resources/redisson_config.yaml"));
        RedissonClient client = Redisson.create(config);
        RTopic topic = client.getTopic("tmiReceiver");
        topic.addListener(String.class, (channel, twitchChannel) -> reader.joinChannel(twitchChannel));
    }
}

File: src/main/java/de/com/fdm/core/RedisListener.kt
package de.com.fdm.core

import org.redisson.Redisson
import org.redisson.api.RedissonClient
import org.redisson.config.Config
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.io.File

//TODO: offload channels if load gets too high over time
@Service
class RedisListener @Autowired constructor(
    private val loadBalancer: LoadBalancer,
) {
    private val log = LoggerFactory.getLogger(RedisListener::class.java)
    private val client: RedissonClient

    init {
        val config = Config.fromYAML(File("src/main/resources/redisson_config.yaml"))
        this.client = Redisson.create(config)
    }

    fun start() {
        val topic = client.getTopic("tmiReceiver")
        topic.addListener(String::class.java) { _: CharSequence, channel: String ->
            loadBalancer.joinChannel(channel)
        }
    }
}
